### PR TITLE
configure.ac: fix the line `$cc_shlib_conf`

### PR DESCRIPTION
**Pull request by hzhou**
Fix issue #3730.

### CODE REVIEW

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
I don't think this change is correct.  `cc_shlib_conf` needs to be executed so that `hardcode_libdir_flag_spec` can be set correctly.  I probably should have added a comment about it when I wrote the code (sorry about that).

**Comment by pavanbalaji** [comment]
> So `$cc_shlib_conf` is a user shell script that supposed to be run (or sourced)? -- That makes sense. Issue #3730 is reporting a permission issue. Maybe we should do `source $cc_shlib_conf`?

That might work.  Worth a shot.  Can you please also add a comment?

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
I am still trying to understand why sourcing `cc_shlib_conf` breaks the build (due to dropping dynamic library target in libtool). However, as for the original reason -- setting `WRAPPER_C_DYNAMIC_LOADING_FLAGS` -- I don't find anywhere this variable is needed. The `mpicc-` family directly uses `AC_SUBST_FILE([cc_shlib_conf])`.

https://github.com/pmodels/mpich/blob/01774e6487222706349617971d6c50894cec9195/configure.ac#L2065-L2081

**Comment by pavanbalaji** [comment]
> I am still trying to understand why sourcing `cc_shlib_conf` breaks the build (due to dropping dynamic library target in libtool). However, as for the original reason -- setting `WRAPPER_C_DYNAMIC_LOADING_FLAGS` -- I couldn't find anywhere this variable is needed. The `mpicc-` family directly uses `AC_SUBST_FILE([cc_shlib_conf])`.

It's only used on `pkgconfig`.

```
% git grep WRAPPER_C_DYNAMIC_LOADING_FLAGS
configure.ac:   eval WRAPPER_C_DYNAMIC_LOADING_FLAGS=\"${hardcode_libdir_flag_spec} ${enable_dtags_flag}\"
configure.ac:   eval WRAPPER_C_DYNAMIC_LOADING_FLAGS=\"${hardcode_libdir_flag_spec} ${disable_dtags_flag}\"
configure.ac:   WRAPPER_C_DYNAMIC_LOADING_FLAGS=""
configure.ac:AC_SUBST(WRAPPER_C_DYNAMIC_LOADING_FLAGS)
src/packaging/pkgconfig/mpich.pc.in:Libs: @WRAPPER_C_DYNAMIC_LOADING_FLAGS@ @WRAPPER_LDFLAGS@ -L${libdir} -l@MPILIBNAME@ @LPMPILIBNAME@ @WRAPPER_LIBS@
```

**Comment by hzhou** [comment]
`cc_shlib.conf` overwrites `libtool`'s `library_names_spec`:
[diff libtool]
```
259c259
< library_names_spec="\$libname\$release\$shared_ext\$versuffix \$libname\$release\$shared_ext\$major \$libname\$shared_ext"
---
> library_names_spec="\$libname\$shrext"

```

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
@pavanbalaji Now the fix seems working. Could you give another review?

PS: I think the way we deal with shared library is too messy and could use an overhaul.

File: configure.ac [hunk 1]
```diff
@@ -2064,7 +2064,14 @@ cc_shlib_conf=src/env/cc_shlib.conf
...
 
-$cc_shlib_conf
+# We need variables from $cc_shlib_conf to set `WRAPPER_C_DYNAMIC_LOADING_FLAGS`, which 
+#    is used in `src/packaging/pkgconfig/mpich.pc.in`
+# However, libtool's $library_names_spec is more than what's given in $cc_shlib_conf, so
+#    we need protect it. 
+# FIXME: potentially, there are more variable collisions. Need better plan.
+save_library_names_spec="$library_names_spec"
+. $cc_shlib_conf
+library_names_spec="$save_library_names_spec"
```
Comment (pavanbalaji) [inline configure.ac]: Can you use the PUSH/POP macros for saving variables?

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Pull request merged**